crates/conjure_core/src/rule_engine/rewrite_naive.rs

```diff
@@ -0,0 +1,84 @@
+use std::sync::Arc;
+
+use uniplate::Biplate;
+
+use super::{RewriteError, RuleSet};
+use crate::{
+    ast::{pretty::pretty_expressions_as_conjunction, Expression as Expr},
+    bug,
+    rule_engine::{get_rule_priorities, get_rules_vec, Reduction},
+    Model,
+};
+
+/// A naive, exhaustive rewriter.
+///
+/// **This rewriter should not be used in production, and is intended as a development tool.**
+///
+/// The goal of this rewriter is to model the correct rule application order. To this end, it uses
+/// the simplest implementation possible, disregarding performance.
+///
+/// **Rule application order:** apply the highest priority rule possible anywhere in the tree,
+/// favouring larger expressions as a tie-breaker.
+
+pub fn rewrite_naive<'a>(
+    model: &Model,
+    rule_sets: &Vec<&'a RuleSet<'a>>,
+) -> Result<Model, RewriteError> {
+    // At each iteration, rules are checked against all expressions in order of priority until one
+    // is applicable. This is done until no more rules can be applied to any expression.
+
+    // rules sorted by priority
+    let rules = get_rules_vec(
+        &get_rule_priorities(rule_sets).unwrap_or_else(|_| bug!("get_rule_priorities() failed!")),
+    );
+
+    type CtxFn = Arc<dyn Fn(Expr) -> Vec<Expr>>;
+
+    // result of the pass, including some logging information
+    // reduction,rule name, priority, original expression, context fn
+    let mut result: Option<(Reduction, String, u16, Expr, CtxFn)>;
+
+    let mut model = model.clone();
+
+    loop {
+        result = None;
+
+        'rules: for rule in &rules {
```

Comment:
this looks very good, though I have one concern.

you are not grouping the rules, which still means you are applying a highest priority rule, at the right spot, so not a problem.

the only thing we are missing out on with this is an opportunity to do rule selection, or indeed an assertion that only on rule of the same priority is applicable.

I think this can be useful for debugging the rules and should be a simple enough change?

instead of rule-priority-sorted vec of rules, we should have a vec of set of rules. then we try everyhing in a particular set, if none are applicable move on, if one is applicable apply, if multiple things are applicable assert.

the assert is where a rule selection function can be inserted in the future.

Reply:
We only want to error when two rules of equal priority are applicable to the same expression, right? Tracking this with this implementation will be annoying... This was easier with the previous attempt, where I iterated over expressions first then rules, so going back to that might be good if keeping. track of all the rules for a given expression is a requirement.

What I can do with this implementation is log a list of all applicable rules across the tree in order of appearance (i.e. by rule priority then position in tree), which for debugging purposes may be fine?

Reply:
I have replicated the `bug!` call for equally applicable rules.

Do you think we need the warning for when rules of different priorities are applicable to an expression here?

Reply:
oh good. I was going to take a look and see why it's annoying, but you beat me to it.

the warning would also be good I suppose, but not a blocker.

I'll review the new change.